azurerm/internal/services/containers/container_group_resource.go

```diff
@@ -177,6 +177,36 @@ func resourceContainerGroup() *schema.Resource {
 				ForceNew: true,
 			},
 
+			"exposed_port": {
+				Type:       schema.TypeSet,
+				Optional:   true, // change to 'Required' in 3.0 of the provider
+				ForceNew:   true,
+				Computed:   true,                        // remove in 3.0 of the provider
+				ConfigMode: schema.SchemaConfigModeAttr, // remove in 3.0 of the provider
```

Comment:
With both optional and computed set to true, I believe there's no way to explicitly force removal of all exposed_port blocks. Removing the last exposed_port block from terraform config doesn't change the container instance at all.

See: https://www.terraform.io/docs/extend/terraform-0.12-compatibility.html#computed-resource-attributes
(and I guess also [this comment in schema.go](https://github.com/hashicorp/terraform/blob/6697245c18dad6848e7647598dda1ab04c2680ca/internal/legacy/helper/schema/schema.go#L78))

With `ConfigMode: schema.SchemaConfigModeAttr`, removing the last exposed_port block still doesn't change the container instance config. But at least we can explicitly set `exposed_port = []`, which allows the user to go back to the default/old behavior of all ports on all containers defined in the container blocks being exposed.

Once we deprecate the default/old behavior, there should be no need to set `exposed_port = []` since it will be a required attribute at that point. So at that point I think we should be able to safely remove `ConfigMode: schema.SchemaConfigModeAttr`.

That was my reasoning at least. Maybe we don't need to set `exposed_port = []`, and maybe the `schema.SchemaConfigModeAttr` is just adding complexity. But the behavior when removing all the exposed_port blocks confused me while I was testing these changes, so I wanted to provide a way to go back to the default/old behavior.

Reply:
So if you set `exposed_port = []`, it does switch back to the default/old behavior. For example, if I have two ports exposed: 443, and 80 then removing those and setting `exposed_port = []` results in a plan like this:

```hcl
      ~ exposed_port        = [ # forces replacement
          - {
              - port     = 443
              - protocol = "TCP"
            },
          - {
              - port     = 80
              - protocol = "TCP"
            },
```

Applying the plan results in the old behavior of all individual container ports being exposed.

But I guess one weird thing is if I run again (without any Terraform config changes, so still with `exposed_port = []`), I get a plan that looks like this:

```hcl
      ~ exposed_port        = [ # forces replacement
          - {
              - port     = 4141
              - protocol = "TCP"
            },
          - {
              - port     = 4180
              - protocol = "TCP"
            },
          - {
              - port     = 443
              - protocol = "TCP"
            },
          - {
              - port     = 6379
              - protocol = "TCP"
            },
          - {
              - port     = 80
              - protocol = "TCP"
            },
          - {
              - port     = 9998
              - protocol = "TCP"
            },
          - {
              - port     = 9999
              - protocol = "TCP"
            },
        ]
```
(all the ports above are exposed on individual containers in my example)

So I guess the schema.SchemaConfigModeAttr is maybe not the cleanest way to deal with this. We could just remove it and put a note in documentation that going back to the old behavior requires tainting the container group?

Sorry in advance if I'm overcomplicating this!

Reply:
If this is a optional and computed list of anything we need config mode attr to sent empty vlaues

Reply:
The issue is I saw no planned changes when going from something like this:

```hcl
resource "azurerm_container_group" "test" {
  name                = "test"

  exposed_port {
    port     = 443
    protocol = "TCP"
  }

  container {
    name   = "testcontainer"

    ports {
      port     = 80
      protocol = "TCP"
    }
    ports {
      port     = 443
      protocol = "TCP"
    }
}
```

To something like this:

```hcl
resource "azurerm_container_group" "test" {
  name                = "test"

  container {
    name   = "testcontainer"

    ports {
      port     = 80
      protocol = "TCP"
    }
    ports {
      port     = 443
      protocol = "TCP"
    }
}
```

So no opportunity to force new. The 2nd config implies that both 80 and 443 should be exposed on the group level. But if you deployed the 1st config and then tried to change to the 2nd one later on, then your config wouldn't match the real infrastructure (which would still be exposing port 443 only).

I thought being able to set `exposed_port = []` would be a good way to work around this. But tainting the resource would also achieve the same results (since both ways would force a new resource).

Reply:
If using ForceNew alone doesn't cause Terraform to recognise an empty list, then I think it's OK to use SchemaConfigModeAttr at least until 3.0 when we can make this a required field. I"m happy to resolve this one.